src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/Descriptors.cs

```diff
@@ -172,92 +172,99 @@ internal class Descriptors
+
+        public static DiagnosticDescriptor AZC0106 = new DiagnosticDescriptor(
+            nameof(AZC0106),
+            "Non-public asynchronous method needs 'async' parameter.",
+            "Non-public asynchronous method that is called in synchronous scope should have a boolean 'async' parameter.",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0107 = new DiagnosticDescriptor(
+            nameof(AZC0107),
+            "DO NOT call public asynchronous method in synchronous scope.",
+            "Public asynchronous method shouldn't be called in synchronous scope. Use synchronous version of the method if it is available.",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0108 = new DiagnosticDescriptor(
+            nameof(AZC0108),
+            "Incorrect 'async' parameter value.",
+            "In {0} scope 'async' parameter for the '{1}' method call should {2}.",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0109 = new DiagnosticDescriptor(
+            nameof(AZC0109),
+            "Misuse of 'async' parameter.",
+            "'async' parameter in asynchronous method can't be changed and can only be used as an exclusive condition in '?:' operator or conditional statement.",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0110 = new DiagnosticDescriptor(
+            nameof(AZC0110),
+            "DO NOT use await keyword in possibly synchronous scope.",
+            "Asynchronous method with `async` parameter can be called from both synchronous and asynchronous scopes. 'await' keyword can be safely used either in guaranteed asynchronous scope (i.e. `if (async) {...}`) or if `async` parameter is passed into awaited method. Awaiting on variables, fields, properties, conditional operators or async methods that don't use `async` parameter isn't allowed outside of the guaranteed asynchronous scope.",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0111 = new DiagnosticDescriptor(
+            nameof(AZC0111),
+            "DO NOT use EnsureCompleted in possibly asynchronous scope.",
+            "Asynchronous method with `async` parameter can be called from both synchronous and asynchronous scopes. 'EnsureCompleted' extension method can be safely used on in guaranteed synchronous scope (i.e. `if (!async) {...}`).",
+            DiagnosticCategory.Usage,
+            DiagnosticSeverity.Warning, true);
+
+        public static DiagnosticDescriptor AZC0112 = new DiagnosticDescriptor(
+            "AZC0112",
+            "Misuse of internal type via InternalsVisibleTo.",
```

Comment:
Technically,
```suggestion
            "Misuse of internal type via InternalsVisibleToAttribute.",
```

Or the attribute syntax like you did in the line below. I'd keep it consistent. But a nit, really.